src/freertos_drivers/common/DccOutput.hxx

```diff
@@ -0,0 +1,364 @@
+    /// Default instance to be used.
+    static const DccOutputImpl<HW> instance_;
+};
+
+/// Allocates the storage by the linker for the static default instance.
+template <class HW> const DccOutputImpl<HW> DccOutputImpl<HW>::instance_;
+
+/// This structure represents a single output channel for a DCC command
+/// station. The DCC driver uses these structures, with the business logic
+/// filled in by the hardware implementor.
+///
+/// @param OUTPUT_NUMBER 0,1,... the number of outputs. Each output is
+/// independently controlled.
+template <int OUTPUT_NUMBER> struct DccOutputHw
+{
+public:
+    /// Bitmask of why this output should be disabled. If zero, the output
+    /// should be enabled. If non-zero, the output should be disabled. A
+    /// variety of system components own one bit in this bitmask each; see {
+    /// \link DccOutput::DisableReason } These bits are all set by the
+    /// application. The DCC Driver will only read this variable, and enable
+    /// the output if all bits are zero.
+    static std::atomic_uint8_t outputDisableReasons_;
+
+    /// 0 if we should not produce a railcom cutout; 1 for short cutout; 2 for
+    /// regular cutout. Set by the application and read by the DCC driver.
+    static std::atomic_uint8_t isRailcomCutoutEnabled_;
+
+    /// 1 if we are in a railcom cutout currently. Set and cleared by the
+    /// driver before calling the start/stop railcom cutout functions.
+    static std::atomic_uint8_t isRailcomCutoutActive_;
+
+    /// Called by the driver to decide whether to make this channel participate
+    /// in the railcom cutout.
+    static bool need_railcom_cutout()
+    {
+        return (outputDisableReasons_ == 0) && (isRailcomCutoutEnabled_ != 0);
+    }
+
+    /// Called once after the railcom cutout is done to decide whether this
+    /// output should be reenabled.
+    static bool should_be_enabled()
+    {
+        return outputDisableReasons_ == 0;
+    }
+
+    /// Clears a disable reason. If all disable reasons are clear, the output
+    /// will be enabled by the DCC driver at the beginning of the next packet.
+    static void clear_disable_reason(DccOutput::DisableReason bit)
+    {
+        outputDisableReasons_ &= ~((uint8_t)bit);
+    }
+
+protected:
+    /// Set one bit in the disable reasons bit field. This is protected,
+    /// because the implementation / child class should have a composite
+    /// function that sets the bit and disables the output in one call.
+    static void set_disable_reason_impl(DccOutput::DisableReason bit)
+    {
+        outputDisableReasons_ |= (uint8_t)bit;
+    }
+
+private:
+    /// Private constructor. These objects cannot be initialized and must only
+    /// have static members.
+    DccOutputHw();
+};
+
+template <int N>
+std::atomic_uint8_t DccOutputHw<N>::outputDisableReasons_ {
+    (uint8_t)DccOutput::DisableReason::INITIALIZATION_PENDING};
+template <int N>
+std::atomic_uint8_t DccOutputHw<N>::isRailcomCutoutEnabled_ {2};
```

Comment:
any reason not to use the enum const here?

Reply:
Done